### PR TITLE
SA-2543: Parallel Implementation

### PR DESCRIPTION
**Pull request by gweinjc**
## Issues
* [SA-2543](https://jumpcloud.atlassian.net/browse/SA-2543) - Use Get-JCResults in applicable Get-JC* public functions to make API calls

## What does this solve?
This will allow users to make use of PowerShell 7+ Parallel functionality natively within the module for the following functions:
* **Get-JCUser**
* **Get-JCSystem**
* **Get-JCSystemUser**
* **Get-JCSystemGroupMember**
* **Get-JCUserGroupMember** 

This is particularly beneficial for customers that are querying large quantities of data. Making use of parallel processing will potentially cause functions to complete **_up to_** 60% faster. As an experiment, I measured the time to complete for each of the functions mentioned above in parallel vs in serial. 

The organization that the tests were performed on has excess of 4,000 users and over 20,000 devices.
<p align="center">

> :bulb: **Tip:** For more fun graphs, hit the drop down below!

<details>
<summary> Individual function comparisons </summary>

### Get-JCUser
<p align="center">

</p>

### Get-JCSystem
<p align="center">

</p>

### Get-JCSystemUser
<p align="center">

</p>

### Get-JCSystemGroupMember
<p align="center">

</p>

<p align="center">

</p>
</details>
</p>

## Is there anything particularly tricky?
There are a few limitations with running these functions in parallel. The two basic ones are the following:
* PowerShell version 7+ installed
* Device running module has multiple cores

> :warning: **Warning:** If those two requirements are not met, the functions will not run in parallel or will not reduce the time to completion as much as anticipated.

Another consideration that needs to be taken into account is the size of the organization and which endpoints are being called. Here are the functions requirements for amount of users/devices in order to successfully run in parallel:
* Get-JCUser: 1001+ Users
    * [Using /search/ endpoint](https://docs.jumpcloud.com/api/1.0/index.html#operation/search_systemusers_post)
* Get-JCSystem: 1001+ Systems
    * [Using /search/ endpoint](https://docs.jumpcloud.com/api/1.0/index.html#operation/search_system_post)
* Get-JCSystemUser: 101+ Users
    * [Using /systems/{system_id}/users endpoint](https://docs.jumpcloud.com/api/2.0/index.html#operation/graph_systemTraverseUser)
* Get-JCSystemGroupMember: 101+ Systems
    * [Using /Systemgroups/{group_id}/members endpoint](https://docs.jumpcloud.com/api/2.0/index.html#operation/graph_systemGroupMembersList)
* Get-JCUserGroupMember: 101+ Users
    * [Using /Usergroups/{group_id}/members endpoint](https://docs.jumpcloud.com/api/2.0/index.html#operation/graph_userGroupMembersList)

> :memo: **Note:** Even if the -parallel flag is set to $true and the above limits are not met, the function will default to running in serial

## How should this be tested?
Ensure that your testing organization meets the following requirements
* Get-JCUser: More than 1001 users
* Get-JCSystem: More than 1001 systems
* Get-JCSystemUser: More than 101 Users assigned to a device
* Get-JCSystemGroupMember: More than 101 Systems assigned to a System/Device Group
* Get-JCUserGroupMember: More than 101 Users assigned to a User Group

Pull this branch and use the following code example to compare results:
```powershell
Import-Module "/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/JumpCloud.psd1" -force
Connect-JCOnline "APIKEY" -force
#Parallel test
Write-Host "Starting parallel..."
$ParallelUsers = Get-JCUser -parallel $true
$ParallelSystems = Get-JCSystem -parallel $true
$ParallelSystemUsers = Get-JCSystemUser -SystemID "SYSTEMID" -parallel $true
$ParallelSystemGroupMembers = Get-JCSystemGroupMember -ByID 'SYSTEMGROUPID' -parallel $true
$ParallelUserGroupMembers = Get-JCUserGroupMember -ByID "USERGROUPID" -parallel $true
Write-Host "Parallel done..."

#Serial test
Write-Host "Starting serial..."
$SerialUsers = Get-JCUser
$SerialSystems = Get-JCSystem
$SerialSystemUsers = Get-JCSystemUser -SystemID "SYSTEMID"
$SerialSystemGroupMembers = Get-JCSystemGroupMember -ByID 'SYSTEMGROUPID'
$SerialUserGroupMembers = Get-JCUserGroupMember -ByID "USERGROUPID"
Write-Host "Serial done..."

#Compare results
$ParallelUsers.Count | Should -Be $SerialUsers.Count
$SortedParallelUsers = $ParallelUsers.username | Sort-Object
$SortedSerialUsers = $SerialUsers.username | Sort-Object
$SortedParallelUsers | Should -Be $SortedSerialUsers

$ParallelSystems.Count | Should -Be $SerialSystems.Count
$SortedParallelSystems = $ParallelSystems.displayName | Sort-Object
$SortedSerialSystems = $SerialSystems.displayName | Sort-Object
$SortedParallelSystems | Should -Be $SortedSerialSystems

$ParallelSystemUsers.Count | Should -Be $SerialSystemUsers.Count
$SortedParallelSysUsers = $ParallelSystemUsers.username | Sort-Object
$SortedSerialSysUsers = $SerialSystemUsers.username | Sort-Object
$SortedParallelSysUsers | Should -Be $SortedSerialSysUsers

$ParallelSystemGroupMembers.Count | Should -Be $SerialSystemGroupMembers.Count
$SortedParallelMembers = $ParallelSystemGroupMembers | Sort-Object -Property SystemID | ConvertTo-Json
$SortedSerialMembers = $SerialSystemGroupMembers | Sort-Object -Property SystemID | ConvertTo-Json
$SortedParallelMembers | Should -Be $SortedSerialMembers

$ParallelUserGroupMembers.Count | Should -Be $SerialUserGroupMembers.Count
$SortedParallelMembers = $ParallelUserGroupMembers | Sort-Object -Property UserID | ConvertTo-Json
$SortedSerialMembers = $SerialUserGroupMembers | Sort-Object -Property UserID | ConvertTo-Json
$SortedParallelMembers | Should -Be $SortedSerialMembers
```

### CODE REVIEW

**Review by kmaranionjc — APPROVED** [review]

File: PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1 [hunk 1]
```diff
@@ -46,28 +46,57 @@ Function Get-JCResults
...
         catch {
             throw $_
         }
 
         # Find total amount of results
-        $totalCount = $response.Headers."x-total-count"
-        $totalCount = [int]$totalCount.Trim()
-        Write-Debug "total count: $totalCount"
-
+        if ($null -eq $response.Headers."x-total-count") {
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1:64]: This if/else statement is the difference between /API/Search and /API/systemUsers endpoints (just as an example right?) We should add a comment here.

File: PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1 [hunk 2]
```diff
@@ -147,6 +223,7 @@ Function Get-JCResults
                 }
                 else {
                     try {
+                        $ProgressPreference = "SilentlyContinue"
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1:226]: What is the scope on ProgressPreference? Is it Per Function/ Per Session? I'm just wondering if we are setting it, do we need to un-set it in the end block of this function?

**Comment by jworkmanjc** [comment]
Tests passed, nice work on this seriously. Graphs are awesome and the PR looks really good. Let's consider a card for the parallel validation function in a future release, I'll get a card together.

File: PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1 (hunk 2 above)
Comment (gweinjc) [reply PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1:226]: https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7.2#progresspreference

It sounds like it's for the function, however I'm not 100% sure. It's not a functionality change compared to the default value other than removing any visual progress indication. It was particularly noticeable with successive API calls, and would actually cause some strange artifacting in the output (random linebreaks/whitespaces/etc)

File: PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1 (hunk 1 above)
Comment (gweinjc) [reply PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1:64]: Agreed, I thought I had added that comment but might have accidentally deleted it at some point, will re-add

File: PowerShell/JumpCloud Module/Public/Groups/SystemGroups/Get-JCSystemGroupMember.ps1 [hunk 3]
```diff
@@ -26,18 +18,28 @@ Function Get-JCSystemGroupMember ()
         Write-Debug 'Verifying JCAPI Key'
         if ($JCAPIKEY.length -ne 40) {Connect-JConline}
 
-        Write-Debug 'Initilizing resultsArray and results ArraryByID'
-        $rawResults = @()
-        $resultsArray = @()
+        if (($PSVersionTable.PSVersion.Major -ge 7) -and ($parallel -eq $true)) {
```
[the remark this answers is not in the document]
Comment (gweinjc) [reply PowerShell/JumpCloud Module/Public/Groups/SystemGroups/Get-JCSystemGroupMember.ps1:21]: Agreed, will write a card

[the remark this answers is not in the document]
Comment (gweinjc) [reply PowerShell/JumpCloud Module/Public/Groups/SystemGroups/Get-JCSystemGroupMember.ps1:21]: Card written: https://jumpcloud.atlassian.net/browse/SA-2696

**Pull request merged**